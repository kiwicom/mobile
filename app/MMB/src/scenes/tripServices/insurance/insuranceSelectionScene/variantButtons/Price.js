// @flow

import * as React from 'react';
import {
  Price as OriginalPrice,
  type StylePropType,
  Text,
  StyleSheet,
} from '@kiwicom/mobile-shared';
import { defaultTokens } from '@kiwicom/mobile-orbit';
import { Translation } from '@kiwicom/mobile-localization';

type PriceType = {|
  +amount: ?number,
  +currency: ?string,
|};

type Props = {|
  +price: ?PriceType,
  +isSelected: boolean,
  +textDisabled: StylePropType,
|};

const Price = (props: Props) => {
  const { price, isSelected, textDisabled } = props;
  const style = isSelected ? styles.price : [styles.price, textDisabled];
  if (price === null) {
    return (
      <Text style={style}>
        <Translation id="mmb.trip_services.insurance.selection.price.free" />
      </Text>
    );
  } else if (price && price.amount != null && price.currency != null) {
    return <OriginalPrice price={price} style={style} />;
  }
  return (
    <Text style={style}>
      <Translation id="mmb.trip_services.insurance.selection.price.not_available" />
    </Text>
  );
};

const styles = StyleSheet.create({
  price: {
    color: defaultTokens.paletteInkNormal,
    fontSize: 13,
  },
});

export default Price;
