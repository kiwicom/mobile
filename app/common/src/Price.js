// @flow

import * as React from 'react';
import { Text } from 'react-native';

type Props = {|
  amount: number | null, // may be null (compatible with GraphQL query failure)
  currency: string | null, // may be null (compatible with GraphQL query failure)
  // $FlowFixMeProps
  style?: Object,
  // $FlowFixMeProps
  amountStyle?: Object,
  // $FlowFixMeProps
  currencyStyle?: Object,
|};

/**
 * Currently we are not doing proper formatting because we do not have enough
 * information about currencies being before/after the amount number. However,
 * all prices should be wrapped in this component so the future changes are easy.
 */
export default function Price(props: Props) {
  return (
    <Text style={props.style}>
      <Text style={props.amountStyle}>{props.amount}</Text>{' '}
      <Text style={props.currencyStyle}>{props.currency}</Text>
    </Text>
  );
}
