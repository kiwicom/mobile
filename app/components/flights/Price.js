// @flow

import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { Text } from 'react-native';

import type { Price } from './__generated__/Price.graphql';

type Props = {
  data: Price,
};

const styles = {
  fontWeight: 'bold',
};

export const PriceWithoutData = ({ data }: Props) => {
  if (!data) {
    return null;
  }
  return (
    <Text style={styles}>
      {data && data.amount} {data && data.currency}
    </Text>
  );
};

export default createFragmentContainer(
  PriceWithoutData,
  graphql`
    fragment Price on Price {
      amount
      currency
    }
  `,
);
