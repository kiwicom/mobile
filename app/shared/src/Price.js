// @flow

import * as React from 'react';
import { Translation } from '@kiwicom/react-native-app-localization';

import CurrencyFormatter from './currency/CurrencyFormatter';
import Text from './Text';
import type { StylePropType } from '../types/Styles';

type Props = {|
  amount: number | null, // may be null (compatible with GraphQL query failure)
  currency: string | null, // may be null (compatible with GraphQL query failure)
  style?: StylePropType,
|};

/**
 * Currently we are not doing proper formatting because we do not have enough
 * information about currencies being before/after the amount number. However,
 * all prices should be wrapped in this component so the future changes are easy.
 */
export default function Price(props: Props) {
  const amount =
    props.amount != null && props.currency != null
      ? CurrencyFormatter(props.amount, props.currency)
      : '';

  return (
    <Text style={props.style}>
      <Translation passThrough={amount} />
    </Text>
  );
}
