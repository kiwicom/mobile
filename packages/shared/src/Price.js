// @flow

import * as React from 'react';
import { Translation } from '@kiwicom/mobile-localization';

import CurrencyFormatter from './CurrencyFormatter';
import Text from './Text';
import type { StylePropType } from '../types/Styles';
import CancellablePromise, {
  type CancellablePromiseType,
} from './CancellablePromise';

type PriceType = {
  +amount?: ?number,
  +currency?: ?string,
};

type Props = {|
  +price?: ?PriceType,
  +style?: StylePropType,
|};

type State = {|
  formattedCurrency: string,
|};

/**
 * Currently we are not doing proper formatting because we do not have enough
 * information about currencies being before/after the amount number. However,
 * all prices should be wrapped in this component so the future changes are easy.
 */
export default class Price extends React.Component<Props, State> {
  cancellablePromise: CancellablePromiseType | null = null;
  state = {
    formattedCurrency: '',
  };

  componentDidMount = () => {
    this.formatCurrency();
  };

  componentDidUpdate = (prevProps: Props) => {
    if (
      (prevProps.price && prevProps.price.amount) !==
      (this.props.price && this.props.price.amount)
    ) {
      this.formatCurrency();
    }
  };

  componentWillUnmount = () => {
    if (this.cancellablePromise !== null) {
      this.cancellablePromise.cancel();
    }
  };

  formatCurrency = async () => {
    if (
      this.props.price != null &&
      this.props.price.amount != null &&
      this.props.price.currency != null
    ) {
      try {
        this.cancellablePromise = CancellablePromise(
          CurrencyFormatter(this.props.price.amount, this.props.price.currency),
        );
        const formattedCurrency = await this.cancellablePromise.promise;
        this.setState({ formattedCurrency });
        this.cancellablePromise = null;
      } catch (err) {
        this.cancellablePromise = null;
      }
    }
  };

  render = () => {
    return (
      <Text style={this.props.style}>
        <Translation passThrough={this.state.formattedCurrency} />
      </Text>
    );
  };
}
