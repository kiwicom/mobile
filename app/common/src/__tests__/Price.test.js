// @flow

import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Price from '../Price';

const renderer = new ShallowRenderer();

it('renders zeros correctly', () => {
  expect(
    renderer.render(<Price amount={0} currency="CZK" />),
  ).toMatchSnapshot();
});

it('renders null values correctly', () => {
  expect(
    renderer.render(<Price amount={null} currency={null} />),
  ).toMatchSnapshot();
});
