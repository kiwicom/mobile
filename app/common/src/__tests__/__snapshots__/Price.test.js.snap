// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders null values correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={undefined}
  >
    
  </Text>
   
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={undefined}
  />
</Text>
`;

exports[`renders zeros correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={undefined}
  >
    0
  </Text>
   
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={undefined}
  >
    CZK
  </Text>
</Text>
`;
